File: examples/test-projects/small-projects/test-project-js/counter.js
/**
 * Counter - счётчик с событиями (аналог Qt signals/slots)
 */
'use strict';

const EventEmitter = require('events');

const THRESHOLD = 5;

class Counter extends EventEmitter {
    constructor(initial) {
        super();
        this._value = (initial !== undefined) ? initial : 0;
    }

    value() {
        return this._value;
    }

    setValue(newValue) {
        if (this._value === newValue) {
            return;
        }
        this._value = newValue;
        this.emit('valueChanged', this._value);
        if (this._value >= THRESHOLD) {
            this.emit('thresholdReached', THRESHOLD);
        }
    }

    increment() {
        this.setValue(this._value + 1);
    }

    reset() {
        this.setValue(0);
    }
}

module.exports = Counter;

File: examples/test-projects/small-projects/test-project-java/Counter.java
package testproject;

/** Счётчик с порогом — аналог C++ Counter (сигналы заменены вызовами-заглушками). */
public class Counter {
    private int value;
    private static final int THRESHOLD = 10;

    public Counter(int initial) {
        value = initial;
    }

    public void valueChanged(int v) {}

    public void thresholdReached(int t) {}

    public int getValue() {
        return value;
    }

    public void setValue(int newValue) {
        if (value == newValue) {
            return;
        }
        value = newValue;
        valueChanged(value);
        if (value >= THRESHOLD) {
            thresholdReached(THRESHOLD);
        }
    }

    public void increment() {
        setValue(value + 1);
    }

    public void reset() {
        setValue(0);
    }
}
